sentinel-cluster/sentinel-cluster-server-default/src/main/java/com/alibaba/csp/sentinel/cluster/flow/ConcurrentClusterFlowChecker.java

```diff
@@ -0,0 +1,102 @@
+ * limitations under the License.
+ */
+package com.alibaba.csp.sentinel.cluster.flow;
+
+import com.alibaba.csp.sentinel.cluster.TokenResult;
+import com.alibaba.csp.sentinel.cluster.TokenResultStatus;
+import com.alibaba.csp.sentinel.cluster.flow.rule.ClusterFlowRuleManager;
+import com.alibaba.csp.sentinel.cluster.flow.statistic.concurrent.CurrentConcurrencyManager;
+import com.alibaba.csp.sentinel.cluster.flow.statistic.concurrent.TokenCacheNode;
+import com.alibaba.csp.sentinel.cluster.flow.statistic.concurrent.TokenCacheNodeManager;
+import com.alibaba.csp.sentinel.cluster.server.log.ClusterServerStatLogUtil;
+import com.alibaba.csp.sentinel.log.RecordLog;
+import com.alibaba.csp.sentinel.slots.block.ClusterRuleConstant;
+import com.alibaba.csp.sentinel.slots.block.flow.FlowRule;
+
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * @author yunfeiyanggzq
+ */
+final public class ConcurrentClusterFlowChecker {
+
+    public static double calcGlobalThreshold(FlowRule rule) {
+        double count = rule.getCount();
+        switch (rule.getClusterConfig().getThresholdType()) {
+            case ClusterRuleConstant.FLOW_THRESHOLD_GLOBAL:
+                return count;
+            case ClusterRuleConstant.FLOW_THRESHOLD_AVG_LOCAL:
+            default:
+                int connectedCount = ClusterFlowRuleManager.getConnectedCount(rule.getClusterConfig().getFlowId());
+                return count * connectedCount;
+        }
+    }
+
+    public static TokenResult acquireConcurrentToken(/*@Valid*/ String clientAddress,FlowRule rule, int acquireCount) {
+        long flowId = rule.getClusterConfig().getFlowId();
+        AtomicInteger nowCalls = CurrentConcurrencyManager.get(flowId);
+        if (nowCalls == null) {
+            RecordLog.warn("[ConcurrentClusterFlowChecker] Fail to get nowCalls by flowId<{}>", flowId);
+            return new TokenResult(TokenResultStatus.FAIL);
+        }
+
+        // check before enter the lock to improve the efficiency
+        if (nowCalls.get() + acquireCount > calcGlobalThreshold(rule)) {
+            ClusterServerStatLogUtil.log("concurrent|block|" + flowId, acquireCount);
+            return new TokenResult(TokenResultStatus.BLOCKED);
+        }
+
+        // ensure the atomicity of operations
+        // lock different nowCalls to improve the efficiency
+        synchronized (nowCalls) {
+            // check again whether the request can pass.
+            if (nowCalls.get() + acquireCount > calcGlobalThreshold(rule)) {
+                ClusterServerStatLogUtil.log("concurrent|block|" + flowId, acquireCount);
+                return new TokenResult(TokenResultStatus.BLOCKED);
+            } else {
+                nowCalls.getAndAdd(acquireCount);
+            }
+        }
+        ClusterServerStatLogUtil.log("concurrent|pass|" + flowId, acquireCount);
+        TokenCacheNode node = TokenCacheNode.generateTokenCacheNode(rule, acquireCount, clientAddress);
```

Comment:
We may need to pay attention to the memory and GC footprint when there are large amount of requests coming.

Reply:
the max memory used is decided by rules*maxConcurrency, I think memory used is controllable.every token information is small and the token can be replaced fastly during the process of acquiring and releasing, there may be some young gc in the case you mention, I think it can be acceptable

Reply:
We may provide benchmark of the scenario. Frequent YGC may need optimizing.

Reply:
I will do it, Thanks !

Reply:
![图片](https://user-images.githubusercontent.com/43897640/89361708-87680580-d6fe-11ea-8aa4-5598c2fbb8bb.png)